### DISCUSSION TITLE
Pitch: overload `vote()` in `IVotingStrategy` to allow for casting a single vote and multiple votes

### DISCUSSION THREAD

**Discussion by 0xZakk** [allo-protocol/aips#15] [New Ideas]
**Context**

Our current implementation of [`IVotingStrategy`](https://github.com/allo-protocol/allo-contracts/blob/main/contracts/votingStrategy/IVotingStrategy.sol) allows someone to cast multiple votes at once by providing an array of encoded votes:

```solidity
function vote(bytes[] calldata _encodedVotes, address _voterAddress) external virtual payable;
```
[Source](https://github.com/allo-protocol/allo-contracts/blob/main/contracts/votingStrategy/IVotingStrategy.sol#L54)

This method is called in the `RoundImplementation` contract:

```solidity
  /// @notice Invoked by voter to cast votes
  /// @param encodedVotes encoded vote
  function vote(bytes[] memory encodedVotes) external payable {
    // slither-disable-next-line timestamp
    require(
      roundStartTime <= block.timestamp &&
      block.timestamp <= roundEndTime,
      "Round: Round is not active"
    );

    IVotingStrategy(votingStrategy).vote{value: msg.value}(encodedVotes, msg.sender);
  }
```
[Source](https://github.com/allo-protocol/allo-contracts/blob/main/contracts/round/RoundImplementation.sol#L446-L457)

Not every voting strategy is going to want to accept multiple votes. But this method is required by the interface and is the assumed behavior. 

**Suggested Implementation**

We could overload `vote()` and provide two implementations, one for casting multiple votes and one for casting a single vote:

```solidity
function vote(bytes calldata _encodedVotes, address _voterAddress) external virtual payable;

function vote(bytes[] calldata _encodedVotes, address _voterAddress) external virtual payable;
```

We would have to update the `RoundImplementation` contract and overload the `vote()` method there too:

```solidity
/// @notice Invoked by voter to cast votes
/// @param encodedVotes encoded vote
function vote(bytes memory encodedVotes) external payable {
  // slither-disable-next-line timestamp
  require(
    roundStartTime <= block.timestamp &&
    block.timestamp <= roundEndTime,
    "Round: Round is not active"
  );

  IVotingStrategy(votingStrategy).vote{value: msg.value}(encodedVotes, msg.sender);
}

/// @notice Invoked by voter to cast votes
/// @param encodedVotes encoded vote
function vote(bytes[] memory encodedVotes) external payable {
  // slither-disable-next-line timestamp
  require(
    roundStartTime <= block.timestamp &&
    block.timestamp <= roundEndTime,
    "Round: Round is not active"
  );

  IVotingStrategy(votingStrategy).vote{value: msg.value}(encodedVotes, msg.sender);
}
```

Normally, we would implement an internal `_vote()` method but I think we should avoid doing that here because we want people to be able to structure their votes however they want and I think we want to avoid looping through the votes array here.

**Comment by 0xKurt** [comment]
> Not every voting strategy is going to want to accept multiple votes. But this method is required by the interface and is the assumed behavior.

Do you have a concrete example in mind? I wonder how much of an edge case this is and if passing just an array with one vote wouldn't solve it.
